src/MongoDB.Driver.Core/Core/Clusters/DnsMonitor.cs

```diff
@@ -0,0 +1,184 @@
+
+namespace MongoDB.Driver.Core.Clusters
+{
+    internal class DnsMonitor
+    {
+        #region static
+        private static string GetParentDomainName(string domainName)
+        {
+            var index = domainName.IndexOf('.');
+            return domainName.Substring(index + 1);
+        }
+        #endregion
+
+        // private fields
+        private readonly CancellationToken _cancellationToken;
+        private readonly IDnsMonitoringCluster _cluster;
+        private readonly LookupClient _lookupClient;
+        private readonly string _lookupDomainName;
+        private readonly string _parentDomainName;
+        private readonly string _query;
+        private bool _startReturnedNormally; // used for testing
+
+        // constructors
+        public DnsMonitor(IDnsMonitoringCluster cluster, string lookupDomainName, CancellationToken cancellationToken)
+        {
+            _cluster = Ensure.IsNotNull(cluster, nameof(cluster));
+            _lookupDomainName = Ensure.IsNotNull(lookupDomainName, nameof(lookupDomainName));
+            _cancellationToken = cancellationToken;
+            _lookupClient = new LookupClient();
+            _parentDomainName = GetParentDomainName(lookupDomainName);
+            _query = "_mongodb._tcp." + _lookupDomainName;
+        }
+
+        // public properties
+        public bool StartReturnedNormally => _startReturnedNormally;
+
+        // public methods
+        public void Start()
+        {
+            try
+            {
+                Monitor();
+            }
+            catch (OperationCanceledException)
+            {
+                // ignore OperationCanceledException
+            }
+            catch (Exception)
+            {
+                // TODO: log unexpected exceptions
+                throw;
+            }
+            _startReturnedNormally = true;
+        }
+
+        // private methods
+        private TimeSpan ComputeRescanDelay(List<SrvRecord> srvRecords)
+        {
+            var delay = TimeSpan.FromSeconds(60);
+
+            if (srvRecords.Count > 0)
+            {
+                var minTimeToLive = TimeSpan.FromSeconds(srvRecords.Select(s => s.InitialTimeToLive).Min());
+                if (minTimeToLive > delay)
+                {
+                    delay = minTimeToLive;
+                }
+            }
+
+            return delay;
+        }
+
+        private List<DnsEndPoint> GetDnsEndPoints(List<SrvRecord> srvRecords)
+        {
+            var endPoints = new List<DnsEndPoint>();
+
+            foreach (var srvRecord in srvRecords)
+            {
+                var host = srvRecord.Target.ToString();
+                if (host.EndsWith("."))
+                {
+                    host = host.Substring(0, host.Length - 1);
+                }
+
+                if (IsValidHost(host))
+                {
+                    // for testing
+                    if (host == "localhost.mongodb_plus_srv_test.com")
```

Comment:
I should probably remove this.

This allowed me to test a standalone console application connecting to a standalone localhost without actually configuring DNS.